### PR TITLE
add share/lib/python and share/bin to PYTHONPATH & PATH

### LINKED ISSUE #5396: share lib/ and bin/ directories
### Problem

Workflows often build or install libraries and executables at runtime. Where should these be put?

You could put them into the `bin/` and `lib/` directories, but these would be overwritten by reinstallation, so you would have to list every resource you want to self-manage into the `.cylcignore` file which is fiddly.

### Proposed Solution

Add support for a `bin/` and `lib/` directory in the `share/` directory (i.e. put these in job's PATH and PYTHONPATH vars). This would help to colocate files generated at runtime and avoid the reinstall problem.

### PR DESCRIPTION
**Pull request by wxtim**
Closes #5396 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at https://github.com/cylc/cylc-doc/pull/609.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: tests/functional/jobscript/01-share-lib-bin/flow.cylc [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    initial cycle point = 1266
+    [[graph]]
+        R1 = install_cold => run_mypythonscript & run_myscript
+
+[runtime]
+    [[install_cold]]
+        script = """
+            cat > "${CYLC_WORKFLOW_SHARE_DIR}/lib/python/mypythonscript.py" <<__HERE__
+            def my_function():
+                print("I can speak Esparanto like a native.")
```
Comment (hjoliver) [inline tests/functional/jobscript/01-share-lib-bin/flow.cylc:16]:
```suggestion
                print("I can speak Esperanto like a native.")
```
But spell it like I didn't grow up in Esperantium 🤣

File: tests/functional/jobscript/01-share-lib-bin/flow.cylc [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            __HERE__
+
+            cat > "${CYLC_WORKFLOW_SHARE_DIR}/bin/myscript.sh" <<__HERE__
+                #!/usr/bin/bash
+                echo "A sure cure for seasickness is to sit under a tree."
```
Comment (hjoliver) [inline tests/functional/jobscript/01-share-lib-bin/flow.cylc:21]: Or is it?

**Review by hjoliver — APPROVED** [review]
Look good. One trivial spelling fix suggested.

**Comment by hjoliver** [comment]
Functional test failure is for `cylc clean` so I expect the test needs tweaking for the new dirs.

File: tests/functional/jobscript/01-share-lib-bin/flow.cylc (hunk 2 above)
Comment (wxtim) [reply tests/functional/jobscript/01-share-lib-bin/flow.cylc:21]: I feel that one large wave might cure seasickness (and all other illnesses) in a rather permanent way for sailors on _that_.

File: tests/functional/jobscript/01-share-lib-bin/flow.cylc (hunk 1 above)
Comment (wxtim) [reply tests/functional/jobscript/01-share-lib-bin/flow.cylc]: Though given the context it's possible the mis-spelling is deliberate...

**Comment by wxtim** [comment]
> Functional test failure is for `cylc clean` so I expect the test needs tweaking for the new dirs.

I was working on my home PC at that point and couldn't test my fix locally - apparently my ability to imitate the output of `tree` manually isn't perfect...

File: tests/functional/jobscript/01-share-lib-bin/flow.cylc (hunk 2 above)
Comment (oliver-sanders) [reply tests/functional/jobscript/01-share-lib-bin/flow.cylc:21]: @hjoliver that is an insanely niche reference.

Comment (oliver-sanders) [reply tests/functional/jobscript/01-share-lib-bin/flow.cylc:21]: https://www.bbc.co.uk/news/blogs-news-from-elsewhere-35891764

File: cylc/flow/etc/job.sh [hunk 3]
```diff
@@ -132,9 +132,12 @@ cylc__job__main() {
     # System paths:
     # * workflow directory (installed run-dir first).
     export PATH="${CYLC_WORKFLOW_RUN_DIR}/bin:${PATH}"
+    export PATH="${CYLC_WORKFLOW_RUN_DIR}/share/bin:${PATH}"
     export PYTHONPATH="${CYLC_WORKFLOW_RUN_DIR}/lib/python:${PYTHONPATH:-}"
+    export PYTHONPATH="${CYLC_WORKFLOW_RUN_DIR}/share/lib/python:${PYTHONPATH:-}"
```
Comment (MetRonnie) [inline cylc/flow/etc/job.sh:137]: Has there been any discussion on whether `/bin/` or `/share/bin/` should take precedence? With this it is currently `/share/bin/` which feels sensible enough. Worth mentioning in the docs PR?

File: tests/functional/jobscript/01-share-lib-bin.t [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# Workflow creates scripts in CYLC_WORKFLOW_SHARE_DIR in `install_cold` task
+# Which are then used in subsequent tasks. We check for the output of those
+# tasks.
+. "$(dirname "$0")/test_header"
+set_test_number 2
+
+# Setup:
+install_workflow "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
+run_ok "${TEST_NAME_BASE}.validate" cylc validate "${WORKFLOW_NAME}"
+run_ok "${TEST_NAME_BASE}.play" cylc play --no-detach "${WORKFLOW_NAME}"
```
Comment (MetRonnie) [inline tests/functional/jobscript/01-share-lib-bin.t:30]:
```suggestion
workflow_run_ok "${TEST_NAME_BASE}.play" cylc play --no-detach "${WORKFLOW_NAME}"
```

File: cylc/flow/etc/job.sh (hunk 3 above)
Comment (oliver-sanders) [reply cylc/flow/etc/job.sh:137]: Good point. The more "specific" should take precedence, but which is more specific? Probably the one generated later (i.e. the share one) in which case this is correct.

Would be better to use a single `export`, note you can split lines in Bash e.g:

```bash
$ FOO="a\
> b\
> c"
[cylc] $ echo $FOO 
abc
```

Comment (MetRonnie) [reply cylc/flow/etc/job.sh:137]: Still yet to use single imports
```suggestion
    export PATH="${CYLC_WORKFLOW_RUN_DIR}/share/bin:${CYLC_WORKFLOW_RUN_DIR}/bin:${PATH}"
    export PYTHONPATH="${CYLC_WORKFLOW_RUN_DIR}/share/lib/python:${CYLC_WORKFLOW_RUN_DIR}/lib/python:${PYTHONPATH:-}"
```

Comment (MetRonnie) [reply cylc/flow/etc/job.sh:137]: Still yet to use single imports
```suggestion
    export PATH="${CYLC_WORKFLOW_RUN_DIR}/share/bin:${CYLC_WORKFLOW_RUN_DIR}/bin:${PATH}"
    export PYTHONPATH="${CYLC_WORKFLOW_RUN_DIR}/share/lib/python:${CYLC_WORKFLOW_RUN_DIR}/lib/python:${PYTHONPATH:-}"
```

**Review by MetRonnie — APPROVED** [review]

**Review by oliver-sanders — APPROVED** [review]

**Pull request merged**